File: leetcode/p00905/p00905.py
from typing import List

class Solution:
    def sortArrayByParity(self, nums: List[int]) -> List[int]:
        head = 0
        tail = len(nums)-1

        while head < tail:
            if nums[head]%2 == 1 and nums[tail]%2 == 0:
                nums[head], nums[tail] = nums[tail], nums[head]
                head += 1
                tail -= 1
            elif nums[head]%2 == 0:
                head += 1
            elif nums[tail]%2 == 1:
                tail -= 1
        
        return nums

if __name__ == '__main__':
    s = Solution()

    assert s.sortArrayByParity([3,1,2,4]) == [4,2,1,3]
    assert s.sortArrayByParity([0]) == [0]

File: leetcode/p00905/p00905.go
package p00905

// Time Complexity: O(n)
// Space Complexity: O(1)
func sortArrayByParity(nums []int) []int {
	head := 0
	tail := len(nums) - 1

	for head < tail {
		for nums[head]%2 == 0 && head < tail {
			head++
		}
		for nums[tail]%2 == 1 && head < tail {
			tail--
		}

		if head < tail {
			nums[head], nums[tail] = nums[tail], nums[head]
			head++
			tail--
		}
	}

	return nums
}
